src/operators.jl

```diff
@@ -18,63 +18,107 @@
 
 # Number
 # Number--Number obviously already taken care of!
-# Number--AbstractVariableRef
-Base.:+(lhs::Number, rhs::AbstractVariableRef) = GenericAffExpr([rhs],[+1.],convert(Float64,lhs))
-Base.:-(lhs::Number, rhs::AbstractVariableRef) = GenericAffExpr([rhs],[-1.],convert(Float64,lhs))
-Base.:*(lhs::Number, rhs::AbstractVariableRef) = GenericAffExpr([rhs],[convert(Float64,lhs)], 0.)
+# Number--VariableRef
+Base.:+(lhs::Number, rhs::AbstractVariableRef) = GenericAffExpr(convert(Float64, lhs), rhs => 1.0)
+Base.:-(lhs::Number, rhs::AbstractVariableRef) = GenericAffExpr(convert(Float64, lhs), rhs => -1.0)
+Base.:*(lhs::Number, rhs::AbstractVariableRef) = GenericAffExpr(0.0, rhs => convert(Float64,lhs))
 # Number--GenericAffExpr
-Base.:+(lhs::Number, rhs::GenericAffExpr) = GenericAffExpr(copy(rhs.vars),copy(rhs.coeffs),lhs+rhs.constant)
-Base.:-(lhs::Number, rhs::GenericAffExpr) = GenericAffExpr(copy(rhs.vars),    -rhs.coeffs ,lhs-rhs.constant)
-Base.:*(lhs::Number, rhs::GenericAffExpr) = GenericAffExpr(copy(rhs.vars),[lhs*rhs.coeffs[i] for i=1:length(rhs.coeffs)],lhs*rhs.constant)
-# Number--GenericQuadExpr
-Base.:+(lhs::Number, rhs::GenericQuadExpr) = GenericQuadExpr(copy(rhs.qvars1),copy(rhs.qvars2),copy(rhs.qcoeffs),lhs+rhs.aff)
-Base.:-(lhs::Number, rhs::GenericQuadExpr) = GenericQuadExpr(copy(rhs.qvars1),copy(rhs.qvars2),    -rhs.qcoeffs ,lhs-rhs.aff)
-Base.:*(lhs::Number, rhs::GenericQuadExpr) = GenericQuadExpr(copy(rhs.qvars1),copy(rhs.qvars2), lhs*rhs.qcoeffs ,lhs*rhs.aff)
+function Base.:+(lhs::Number, rhs::GenericAffExpr)
+    result = copy(rhs)
+    result.constant += lhs
+    return result
+end
+function Base.:-(lhs::Number, rhs::GenericAffExpr)
+    result = -rhs
+    result.constant += lhs
+    return result
+end
+Base.:*(lhs::Number, rhs::GenericAffExpr) = map_coefficients(rhs, c -> lhs * c)
+# Number--QuadExpr
+Base.:+(lhs::Number, rhs::GenericQuadExpr) = GenericQuadExpr(lhs+rhs.aff, copy(rhs.terms))
+function Base.:-(lhs::Number, rhs::GenericQuadExpr)
+    result = -rhs
+    result.aff.constant += lhs
+    return result
+end
+Base.:*(lhs::Number, rhs::GenericQuadExpr) = map_coefficients(rhs, c -> lhs * c)
 
 # AbstractVariableRef (or, AbstractJuMPScalar)
+# TODO: What is the role of AbstractJuMPScalar??
 Base.:+(lhs::AbstractJuMPScalar) = lhs
-Base.:-(lhs::AbstractVariableRef) = GenericAffExpr([lhs],[-1.0],0.0)
-Base.:*(lhs::AbstractJuMPScalar) = lhs # make this more generic so extensions don't have to define unary multiplication for our macros
+Base.:-(lhs::AbstractVariableRef) = GenericAffExpr(0.0, lhs => -1.0)
+Base.:*(lhs::AbstractVariableRef) = lhs # make this more generic so extensions don't have to define unary multiplication for our macros
```

Comment:
By replacing `AbstractJuMPScalar` with `AbstractVariableRef`, it won't work anymore for `AffExpr` and `QuadExpr`. This should be covered by the tests :/

Reply:
Done and added tests.